File: k9mail/src/main/java/com/fsck/k9/helper/MessageHelper.java
package com.fsck.k9.helper;


import android.content.Context;
import android.support.annotation.VisibleForTesting;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.SpannableStringBuilder;
import android.text.TextUtils;
import android.text.style.ForegroundColorSpan;

import com.fsck.k9.Account;
import com.fsck.k9.K9;
import com.fsck.k9.R;
import com.fsck.k9.activity.FolderInfoHolder;
import com.fsck.k9.activity.MessageInfoHolder;
import com.fsck.k9.mail.Address;
import com.fsck.k9.mail.Flag;
import com.fsck.k9.mail.Message.RecipientType;
import com.fsck.k9.mailstore.LocalMessage;

public class MessageHelper {
    /**
     * If the number of addresses exceeds this value the addresses aren't
     * resolved to the names of Android contacts.
     *
     * <p>
     * TODO: This number was chosen arbitrarily and should be determined by
     * performance tests.
     * </p>
     *
     * @see #toFriendly(Address[], com.fsck.k9.helper.Contacts)
     */
    private static final int TOO_MANY_ADDRESSES = 50;

    private static MessageHelper sInstance;

    public synchronized static MessageHelper getInstance(final Context context) {
        if (sInstance == null) {
            sInstance = new MessageHelper(context);
        }
        return sInstance;
    }

    private Context mContext;

    private MessageHelper(final Context context) {
        mContext = context;
    }

    public void populate(final MessageInfoHolder target,
                         final LocalMessage message,
                         final FolderInfoHolder folder,
                         Account account) {
        final Contacts contactHelper = K9.showContactName() ? Contacts.getInstance(mContext) : null;

        target.message = message;
        target.compareArrival = message.getInternalDate();
        target.compareDate = message.getSentDate();
        if (target.compareDate == null) {
            target.compareDate = message.getInternalDate();
        }

        target.folder = folder;

        target.read = message.isSet(Flag.SEEN);
        target.answered = message.isSet(Flag.ANSWERED);
        target.forwarded = message.isSet(Flag.FORWARDED);
        target.flagged = message.isSet(Flag.FLAGGED);

        Address[] addrs = message.getFrom();

        if (addrs.length > 0 &&  account.isAnIdentity(addrs[0])) {
            CharSequence to = toFriendly(message.getRecipients(RecipientType.TO), contactHelper);
            target.compareCounterparty = to.toString();
            target.sender = new SpannableStringBuilder(mContext.getString(R.string.message_to_label)).append(to);
        } else {
            target.sender = toFriendly(addrs, contactHelper);
            target.compareCounterparty = target.sender.toString();
        }

        if (addrs.length > 0) {
            target.senderAddress = addrs[0].getAddress();
        } else {
            // a reasonable fallback "whomever we were corresponding with
            target.senderAddress = target.compareCounterparty;
        }

        target.uid = message.getUid();
        target.account = message.getFolder().getAccountUuid();
        target.uri = message.getUri();
    }

    public CharSequence getDisplayName(Account account, Address[] fromAddrs, Address[] toAddrs) {
        final Contacts contactHelper = K9.showContactName() ? Contacts.getInstance(mContext) : null;

        CharSequence displayName;
        if (fromAddrs.length > 0 && account.isAnIdentity(fromAddrs[0])) {
            CharSequence to = toFriendly(toAddrs, contactHelper);
            displayName = new SpannableStringBuilder(
                    mContext.getString(R.string.message_to_label)).append(to);
        } else {
            displayName = toFriendly(fromAddrs, contactHelper);
        }

        return displayName;
    }

    public boolean toMe(Account account, Address[] toAddrs) {
        for (Address address : toAddrs) {
            if (account.isAnIdentity(address)) {
                return true;
            }
        }
        return false;
    }

    /**
     * Returns the name of the contact this email address belongs to if
     * the {@link Contacts contacts} parameter is not {@code null} and a
     * contact is found. Otherwise the personal portion of the {@link Address}
     * is returned. If that isn't available either, the email address is
     * returned.
     *
     * @param address An {@link com.fsck.k9.mail.Address}
     * @param contacts A {@link Contacts} instance or {@code null}.
     * @return A "friendly" name for this {@link Address}.
     */
    public static CharSequence toFriendly(Address address, Contacts contacts) {
        return toFriendly(address,contacts,
                K9.showCorrespondentNames(),
                K9.changeContactNameColor(),
                K9.getContactNameColor());
    }

    public static CharSequence toFriendly(Address[] addresses, Contacts contacts) {
        if (addresses == null) {
            return null;
        }

        if (addresses.length >= TOO_MANY_ADDRESSES) {
            // Don't look up contacts if the number of addresses is very high.
            contacts = null;
        }

        SpannableStringBuilder sb = new SpannableStringBuilder();
        for (int i = 0; i < addresses.length; i++) {
            sb.append(toFriendly(addresses[i], contacts));
            if (i < addresses.length - 1) {
                sb.append(',');
            }
        }
        return sb;
    }

    @VisibleForTesting
    /* package, for testing */ static CharSequence toFriendly(Address address, Contacts contacts,
                                                 boolean showCorrespondentNames,
                                                 boolean changeContactNameColor,
                                                 int contactNameColor) {
        if (!showCorrespondentNames) {
            return address.getAddress();
        } else if (contacts != null) {
            final String name = contacts.getNameForAddress(address.getAddress());
            // TODO: The results should probably be cached for performance reasons.
            if (name != null) {
                if (changeContactNameColor) {
                    final SpannableString coloredName = new SpannableString(name);
                    coloredName.setSpan(new ForegroundColorSpan(contactNameColor),
                            0,
                            coloredName.length(),
                            Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
                    );
                    return coloredName;
                } else {
                    return name;
                }
            }
        }

        return (!TextUtils.isEmpty(address.getPersonal())) ? address.getPersonal() : address.getAddress();
    }
}

File: legacy/core/src/main/java/com/fsck/k9/helper/MessageHelper.kt
package com.fsck.k9.helper

import android.text.Spannable
import android.text.SpannableString
import android.text.SpannableStringBuilder
import android.text.TextUtils
import android.text.style.ForegroundColorSpan
import app.k9mail.core.android.common.contact.ContactRepository
import app.k9mail.core.common.mail.toEmailAddressOrNull
import com.fsck.k9.CoreResourceProvider
import com.fsck.k9.K9.contactNameColor
import com.fsck.k9.K9.isChangeContactNameColor
import com.fsck.k9.K9.isShowContactName
import com.fsck.k9.K9.isShowCorrespondentNames
import com.fsck.k9.mail.Address
import java.util.regex.Pattern

class MessageHelper(
    private val resourceProvider: CoreResourceProvider,
    private val contactRepository: ContactRepository,
) {

    fun getSenderDisplayName(address: Address?): CharSequence {
        if (address == null) {
            return resourceProvider.contactUnknownSender()
        }
        val repository = if (isShowContactName) contactRepository else null
        return toFriendly(address, repository)
    }

    fun getRecipientDisplayNames(addresses: Array<Address>?): CharSequence {
        if (addresses == null || addresses.isEmpty()) {
            return resourceProvider.contactUnknownRecipient()
        }
        val repository = if (isShowContactName) contactRepository else null
        val recipients = toFriendly(addresses, repository)
        return SpannableStringBuilder(resourceProvider.contactDisplayNamePrefix()).append(' ').append(recipients)
    }

    companion object {
        /**
         * If the number of addresses exceeds this value the addresses aren't
         * resolved to the names of Android contacts.
         *
         * TODO: This number was chosen arbitrarily and should be determined by performance tests.
         *
         * @see .toFriendly
         */
        private const val TOO_MANY_ADDRESSES = 50
        private val SPOOF_ADDRESS_PATTERN = Pattern.compile("[^(]@")

        /**
         * Returns the name of the contact this email address belongs to if
         * the [contacts][Contacts] parameter is not `null` and a
         * contact is found. Otherwise the personal portion of the [Address]
         * is returned. If that isn't available either, the email address is
         * returned.
         *
         * @param address An [com.fsck.k9.mail.Address]
         * @param contacts A [Contacts] instance or `null`.
         * @return A "friendly" name for this [Address].
         */
        fun toFriendly(address: Address, contactRepository: ContactRepository?): CharSequence {
            return toFriendly(
                address,
                contactRepository,
                isShowCorrespondentNames,
                isChangeContactNameColor,
                contactNameColor,
            )
        }

        fun toFriendly(addresses: Array<Address>?, contactRepository: ContactRepository?): CharSequence? {
            var repository = contactRepository
            if (addresses == null) {
                return null
            }
            if (addresses.size >= TOO_MANY_ADDRESSES) {
                // Don't look up contacts if the number of addresses is very high.
                repository = null
            }
            val stringBuilder = SpannableStringBuilder()
            for (i in addresses.indices) {
                stringBuilder.append(toFriendly(addresses[i], repository))
                if (i < addresses.size - 1) {
                    stringBuilder.append(',')
                }
            }
            return stringBuilder
        }

        /* package, for testing */
        @JvmStatic
        fun toFriendly(
            address: Address,
            contactRepository: ContactRepository?,
            showCorrespondentNames: Boolean,
            changeContactNameColor: Boolean,
            contactNameColor: Int,
        ): CharSequence {
            if (!showCorrespondentNames) {
                return address.address
            } else if (contactRepository != null) {
                val name = contactRepository.getContactName(address)
                if (name != null) {
                    return if (changeContactNameColor) {
                        val coloredName = SpannableString(name)
                        coloredName.setSpan(
                            ForegroundColorSpan(contactNameColor),
                            0,
                            coloredName.length,
                            Spannable.SPAN_EXCLUSIVE_EXCLUSIVE,
                        )
                        coloredName
                    } else {
                        name
                    }
                }
            }
            return if (!TextUtils.isEmpty(address.personal) && !isSpoofAddress(address.personal)) {
                address.personal
            } else {
                address.address
            }
        }

        private fun ContactRepository.getContactName(address: Address): String? {
            return address.address.toEmailAddressOrNull()?.let { emailAddress ->
                getContactFor(emailAddress)?.name
            }
        }

        private fun isSpoofAddress(displayName: String): Boolean {
            return displayName.contains("@") && SPOOF_ADDRESS_PATTERN.matcher(displayName).find()
        }
    }
}
